### DISCUSSION TITLE
How to initialize an AuxiliaryStateVariable in a code aster simulation using MFront material model

### DISCUSSION THREAD

**Discussion by ngrilli** [thelfer/tfel#464] [Q&A] [answered]
I have developed a dislocation-based model for crystal plasticity:

https://github.com/ngrilli/MFrontLaws/blob/main/CrystalPlasticityDislocationUpdate.mfront

it contains the state variable representing the dislocation density:

@AuxiliaryStateVariable strain rho_ssd[Nss];

Normally, in Mtest file:

https://github.com/ngrilli/MFrontLaws/blob/main/CrystalPlasticityDislocationUpdate.mtest

I initialize the dislocation density using the following:

@InternalStateVariable 'rho_ssd' {257.3578,257.3578,257.3578,257.3578,257.3578,257.3578,257.3578,257.3578,257.3578,257.3578,257.3578,257.3578};

How do I initialize the dislocation density in a code aster .comm input file?

I have written the following .comm file representing an 8-grain polycrystal
and I know how to pass the @Parameter using LISTE_COEF,
however, I cannot find an example of initializing an internal state variable in MFront using aster.

If this is not possible or too complicated, is there another option in the .mfront file
to initialize the value of an @AuxiliaryStateVariable ?

I paste below the full Code Aster input file I prepared.

Thank you very much in advance,
Nicolò

DEBUT(LANG='FR')

mesh = LIRE_MAILLAGE(identifier='0:1',
                     FORMAT='MED',
                     UNITE=20);

model = AFFE_MODELE(identifier='1:1',
                    AFFE=_F(MODELISATION='3D',
                            PHENOMENE='MECANIQUE',
                            TOUT='OUI'),
                    MAILLAGE=mesh);

mater = DEFI_MATERIAU(identifier='2:1',
                      ELAS=_F(E=204600.0,NU=0.402),
                      MFRONT=_F(LISTE_COEF=(93812.45,93812.45,93812.45,0.40228,0.40228,0.40228,126200.0,126200.0,126200.0,0.001,0.1,0.0,0.1,0.000256,126200.0,0.3,1.4,0.112,1.13,0.0013,0.0,0.0)));

cara= AFFE_CARA_ELEM(identifier='3:1',
                     MODELE=model, 
                     MASSIF=(_F(GROUP_MA='Grain1', ANGL_EULER=(0.,0.,0.)),
					         _F(GROUP_MA='Grain2', ANGL_EULER=(10.,0.,0.)),
							 _F(GROUP_MA='Grain3', ANGL_EULER=(20.,0.,0.)),
							 _F(GROUP_MA='Grain4', ANGL_EULER=(30.,0.,0.)),
							 _F(GROUP_MA='Grain5', ANGL_EULER=(40.,0.,0.)),
							 _F(GROUP_MA='Grain6', ANGL_EULER=(50.,0.,0.)),
							 _F(GROUP_MA='Grain7', ANGL_EULER=(60.,0.,0.)),
							 _F(GROUP_MA='Grain8', ANGL_EULER=(70.,0.,0.))));
	 
fieldmat = AFFE_MATERIAU(identifier='4:1',
                         AFFE=_F(MATER=(mater, ),
                                 TOUT='OUI'),
                         MODELE=model);				 

listr = DEFI_LIST_REEL(identifier='5:1',
                       DEBUT=0.0,
                       INTERVALLE=_F(JUSQU_A=100.0,
                                     NOMBRE=10000));							 
							 
RampU = DEFI_FONCTION(identifier='6:1',
                      NOM_PARA='INST',
                      VALE=(0.0, 0.0, 100.0, 20.0));

load = AFFE_CHAR_MECA(identifier='7:1',
                      FACE_IMPO=(_F(DX=0.0,
                                    GROUP_MA=('left', )),
                                 _F(DY=0.0,
                                    GROUP_MA=('bottom', )),
                                 _F(DZ=0.0,
                                    GROUP_MA=('back', )),
                                 _F(DX=20.0,
                                    GROUP_MA=('right', ))),
                      MODELE=model);

resnonl = STAT_NON_LINE(identifier='8:1',
                        COMPORTEMENT=_F(DEFORMATION='SIMO_MIEHE',
                                        ITER_INTE_MAXI=1000,
                                        NOM_ROUTINE='astercrystalplasticitydislocationupdate',
                                        RELATION='MFRONT',
                                        LIBRAIRIE='/home/nicolo/projects/SINDRI/MOOSEvsMFront/src/libAsterBehaviour.so'),
                        CHAM_MATER=fieldmat,
                        CARA_ELEM=cara,
						EXCIT=_F(CHARGE=load,
                                 FONC_MULT=RampU),
                        INCREMENT=_F(LIST_INST=listr),
						ARCHIVAGE=_F(PAS_ARCH=1),
                        METHODE='NEWTON',
                        MODELE=model,
                        NEWTON=_F(MATRICE='TANGENTE',
                                  REAC_ITER=1,),
						CONVERGENCE=_F(ITER_GLOB_MAXI=10000,));

resnonl = CALC_CHAMP(identifier='9:1',
                     reuse=resnonl,
                     CONTRAINTE=('SIEF_ELNO', 'SIEF_NOEU', 'SIGM_ELGA', 'SIGM_ELNO', 'SIGM_NOEU'),
                     CRITERES=('SIEQ_ELGA', 'SIEQ_ELNO'),
                     RESULTAT=resnonl);

IMPR_RESU(identifier='10:1',
          FORMAT='MED',
          RESU=_F(RESULTAT=resnonl),
          UNITE=80);

FIN()

**Comment by thelfer** [comment, marked as answer]
@ngrilli I have no clue how to initialize a state variable in `code_aster`, but I know that it's possible. You may want to post a question on the code_aster forum.

In the meantime, you can simply shift the variable by the initial value for instance. I would use a parameter to do that:

```cxx
@Parameter real rho_ssd0 = 1.e-20;
```

Then you can replace each occurence of `rho_ssd` by `rho_ssd+rho_ssd0`...
